---
title: "scRNAseq_enrichment_EDB001"
author: "Daria"
date: "2025-04-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=8, fig.height=6) 
knitr::opts_chunk$set(cache = TRUE, cache.lazy = FALSE)
```

## Setup



```{r libs , message=FALSE, cache = FALSE}
library(stringr)                    
library(dplyr)
library(readxl)
library(clusterProfiler)
#BiocManager::install("org.Hs.eg.db")
organism = "org.Hs.eg.db"
library(organism, character.only = TRUE)
library(enrichR)
library(writexl)
library(ggplot2)
library(ggrepel)
library(ggforce)
library(tidyr)
library(biomaRt)
source("C:/Users/Daria/Documents/R/enrichment_analysis/Enrichment_functions.R")
library(RColorBrewer)
#BiocManager::install("ReactomePA")
library(ReactomePA)
library(viridis)

#library("wesanderson")

theme_set(theme_bw(base_size = 14))
writeLines(capture.output(sessionInfo()), "sessionInfoEnrichment.txt")

OUT_DIR = "enrichment"
dir.create(OUT_DIR)

save.image("enrichment.RData")
```

## Read differential expression data


```{r read_data}

de = readxl::read_excel("all.genes.condition_Tcell.xlsx", sheet = 1)

# change gene names and aliases to Approved.symbol
alias = read.csv("C:/Users/Daria/Documents/ElienDeBousser/HGNC_biomart_gene_aliases.txt", sep = "\t")

alias %>%
  dplyr::filter(Status == "Approved")%>%
  dplyr::select(Approved.symbol, Previous.symbol, Alias.symbol, Ensembl.gene.ID) %>%
  mutate(gene = Approved.symbol)%>%
  dplyr::select(gene, Approved.symbol, Previous.symbol, Alias.symbol, Ensembl.gene.ID) %>%
  pivot_longer(!gene) %>%
  filter(value != "")-> alias2

de$gene = alias2$gene[match(de$gene_name, alias2$value)]
de$Ensembl.gene.ID = alias$Ensembl.gene.ID[match(de$gene, alias$Approved.symbol)]

sum(is.na(de$gene)) #286
sum(is.na(de$Ensembl.gene.ID))

# Convert gene names to Entrez gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
ids<-bitr(de$gene_name, fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
dedup_ids = ids[!duplicated(ids[c("SYMBOL")]),]
de$entrez_gene_id = dedup_ids$ENTREZID[match(de$gene_name, dedup_ids$SYMBOL)]

sum(is.na(de$entrez_gene_id)) #1386 too much!

#use biomart to get EntrezID
listEnsembl()
listEnsemblArchives() #find human ensembl 113
mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl",  host="https://oct2024.archive.ensembl.org")
#listAttributes(mart, what = c("name","description","page"))
biomart_additional=getBM(attributes = c("ensembl_gene_id","entrezgene_id","external_gene_name", "description"),
                         filters="ensembl_gene_id",
                         values=de$Ensembl.gene.ID,
                         mart=mart)
de$entrez_gene_id = biomart_additional$entrezgene_id[match(de$Ensembl.gene.ID, biomart_additional$ensembl_gene_id)]
sum(is.na(de$entrez_gene_id)) # 605

#use bitr to fill in gaps after biomart
de$entrez_gene_id = ifelse(is.na(de$entrez_gene_id), dedup_ids$ENTREZID[match(de$gene_name, dedup_ids$SYMBOL)], de$entrez_gene_id)
sum(is.na(de$entrez_gene_id)) # 322

# filter data that doesn't have approaved gene name
# split by contrast
de %>%
  dplyr::filter(!is.na(gene)) %>%
  dplyr::arrange(desc(avg_log2FC)) %>%
  group_by(contrast) %>%
  group_split() -> res


```

## GSEA

Run GSEA analysis of:
KEGG
WikiPathways
Reactome
MSigDb hallmark 
MSigDb C3 transcription factor targets
Cell Marker

## KEGG

entrez gene id as input

```{r kegg, message=FALSE, warning = FALSE}
#get names of all contrasts
data.frame(number = 1:length(res),
           contrast = sapply(res, function(x) x$contrast[1])
           )
relevant_analyses = c(1:5, 8, 11)
res = res[relevant_analyses]
## KEGG

kgse=list()
  for (i in 1: length(res)){
    kgse[[i]] = gsea_kegg(res[[i]]$avg_log2FC, res[[i]]$entrez_gene_id, res[[i]]$contrast[1], "hsa")
  }


lapply(kgse, function(x) nrow(x[[4]]))

kgse.tbl = list()
for (i in 1:length(kgse)) {kgse.tbl[[i]]=kgse[[i]][[4]] }
write.table( do.call(rbind,kgse.tbl), file=paste0(OUT_DIR,"/output_GSEA_KEGG.tsv"), row.names = FALSE, sep='\t',quote=FALSE )

pdf(file = paste0(OUT_DIR,"/dotplot_GSEA_KEGG.pdf")) 
for (i in 1:length(kgse)){
  if(all(is.na(kgse[[i]][[2]]))){next}
  plot(kgse[[i]][[2]])
}
dev.off()

pdf(file = paste0(OUT_DIR,"/network_GSEA_KEGG.pdf")) 
color <- brewer.pal(9, "RdYlBu")
for (i in 1:length(kgse)){
  if(all(is.na(kgse[[i]][[3]]))){next}
  plot(kgse[[i]][[3]] +
  scale_color_gradient2(high = color[1],low = color[9]) + 
  labs(color = "fold change"))
}
dev.off()




```

## WikiPathways

entrez gene id as input

```{r wiki, message=FALSE, warning = FALSE}

## Wiki

wpgse=list()
  for (i in 1: length(res)){
    wpgse[[i]] = gsea_wiki(res[[i]]$avg_log2FC, res[[i]]$entrez_gene_id, res[[i]]$contrast[1])
  }


lapply(wpgse, function(x) nrow(x[[4]]))

wpgse.tbl = list()
for (i in 1:length(wpgse)) {wpgse.tbl[[i]]=wpgse[[i]][[4]] }
write.table( do.call(rbind,wpgse.tbl), file=paste0(OUT_DIR,"/output_GSEA_WikiPathways.tsv"), row.names = FALSE, sep='\t',quote=FALSE )

pdf(file = paste0(OUT_DIR,"/dotplot_GSEA_WikiPathways.pdf")) 
for (i in 1:length(wpgse)){
  if(all(is.na(wpgse[[i]][[2]]))){next}
  plot(wpgse[[i]][[2]])
}
dev.off()

pdf(file = paste0(OUT_DIR,"/network_GSEA_WikiPathways.pdf")) 
color <- brewer.pal(9, "RdYlBu")
for (i in 1:length(wpgse)){
  if(all(is.na(wpgse[[i]][[3]]))){next}
  plot(wpgse[[i]][[3]] +
  scale_color_gradient2(high = color[1],low = color[9]) + 
  labs(color = "fold change"))
}
dev.off()


```

## Reactome

```{r PC, message=FALSE, warning = FALSE}
#test = gsea_rea(res[[7]]$avg_log2FC, res[[7]]$entrez_gene_id, res[[7]]$contrast[1])

## Reactome

rgse=list()
  for (i in 1: length(res)){
    rgse[[i]] = gsea_rea(res[[i]]$avg_log2FC, res[[i]]$entrez_gene_id, res[[i]]$contrast[1])
  }


lapply(rgse, function(x) nrow(x[[4]]))
lapply(rgse, function(x) colnames(x[[4]]))

rgse.tbl = list()
for (i in 1:length(rgse)) {rgse.tbl[[i]]=rgse[[i]][[4]] }
write.table( do.call(rbind,rgse.tbl), file=paste0(OUT_DIR,"/output_GSEA_Reactome.tsv"), row.names = FALSE, sep='\t',quote=FALSE )

pdf(file = paste0(OUT_DIR,"/dotplot_GSEA_Reactome.pdf")) 
for (i in 1:length(rgse)){
  if(all(is.na(rgse[[i]][[2]]))){next}
  plot(rgse[[i]][[2]])
}
dev.off()

pdf(file = paste0(OUT_DIR,"/network_GSEA_Reactome.pdf")) 
color <- brewer.pal(9, "RdYlBu")
for (i in 1:length(rgse)){
  if(all(is.na(rgse[[i]][[3]]))){next}
  plot(rgse[[i]][[3]] +
  scale_color_gradient2(high = color[1],low = color[9]) + 
  labs(color = "fold change"))
}
dev.off()


```


## overrepresentation analysis

Run overrepresentation analysis using
Enrichr
on all human sets available

```{r}

```

## interesting pathways

Entry	
N00150                      Network                                
Name	
Type I IFN signaling pathway


```{r interesting_pathways, message=FALSE, warning = FALSE}

lapply(res, function(x) x$contrast[1])
color <- brewer.pal(9, "RdYlBu")

pdf(file = paste0(OUT_DIR,"/cell_cycle.pdf")) 
  for (i in 1: length(res)){
    plot = vis_reactome(res[[i]]$avg_log2FC, res[[i]]$entrez_gene_id, res[[i]]$contrast[1], pathway_name = "Cell Cycle")
    print(plot+
            scale_color_gradient2(high = color[1],low = color[9]) +
            labs(color = "fold change"))
    genes_cc = plot$data$name
  }
dev.off()


#viewPathway("Interferon alpha/beta signaling", 
pdf(file = paste0(OUT_DIR,"/interferon_alphabeta_signaling.pdf")) 
  for (i in 1: length(res)){
    plot = vis_reactome(res[[i]]$avg_log2FC, res[[i]]$entrez_gene_id, res[[i]]$contrast[1], 
                        pathway_name = "Interferon alpha/beta signaling")
    print(plot+
            scale_color_gradient2(high = color[1],low = color[9]) +
            labs(color = "fold change"))
    genes_in = plot$data$name
  }
dev.off()
```
## Volcano plots selected genes

```{r}

pdf(file = paste0(OUT_DIR,"/cell_cycle_volcano.pdf")) 
  for (i in 1: length(res)){
# Input table

results = data.frame(
  adj.P.Val = res[[i]]$p_val_adj,
  logFC = res[[i]]$avg_log2FC,
  Gene = res[[i]]$gene,
  P.Val = res[[i]]$p_val
)


# Define thresholds (adjustable by the user)
sig_pval <- 0.05
lfc_threshold <- 0.1

# Allow the user to define an additional list of gene names to highlight (modify as needed)
extra_genes <- genes_cc  # Replace with desired gene names

# Ensure gene names are available in the results
results$Gene <- results$Gene

# Define significance categories based on adjusted p-value and log2 fold change thresholds
results$Significance <- "NotSignificant"
results$Significance[results$adj.P.Val < sig_pval & results$logFC >= lfc_threshold] <- "Upregulated"
results$Significance[results$adj.P.Val < sig_pval & results$logFC <= -lfc_threshold] <- "Downregulated"
results$Significance <- factor(results$Significance, levels = c("Upregulated", "Downregulated", "NotSignificant"))

# Calculate -log10 of adjusted p-values for plotting
results$log10AdjP <- -log10(results$adj.P.Val)
results$log10P <- -log10(results$P.Val)
# Determine symmetrical x-axis limits based on maximum absolute log2 fold change (rounded up)
max_lfc <- ceiling(max(abs(results$logFC), na.rm = TRUE))
x_limits <- c(-max_lfc, max_lfc)

# Identify top 10 upregulated and top 10 downregulated proteins
top_up <- results %>% 
  dplyr::filter(Significance == "Upregulated") %>% 
  dplyr::arrange(-logFC) %>% 
  head(10)
top_down <- results %>% 
  dplyr::filter(Significance == "Downregulated") %>% 
  dplyr::arrange(logFC) %>% 
  head(10)

# Additional genes to highlight (if present in the results)
extra_highlight <- results %>% 
  dplyr::filter(Gene %in% extra_genes)

# Combine labels (avoid duplicates)
#labels_df <- dplyr::bind_rows(top_up, top_down, extra_highlight) %>% 
labels_df <- dplyr::bind_rows(extra_highlight) %>% 
  dplyr::distinct(Gene, .keep_all = TRUE)

# Define colors using the viridis palette for upregulated and downregulated proteins
up_color <- viridis(2)[2]
down_color <- viridis(2)[1]

# Generate the volcano plot
#ggplot(results, aes(x = logFC, y = log10AdjP)) +
plot = ggplot(results, aes(x = logFC, y = log10P)) +
  geom_point(aes(color = Significance), alpha = 0.6) +
  scale_color_manual(values = c("Upregulated" = up_color, 
                                "Downregulated" = down_color, 
                                "NotSignificant" = "grey50")) +
  geom_vline(xintercept = c(-lfc_threshold, lfc_threshold), linetype = "dashed", color = "red") +
  #geom_hline(yintercept = -log10(sig_pval), linetype = "dashed", color = "red") +
  geom_text_repel(data = labels_df, 
                  aes(label = Gene),
                  size = 3,
                  max.overlaps = Inf,
                  min.segment.length = 0
                  ) +
  coord_cartesian(xlim = x_limits) +
  theme_bw() +
  labs(title = paste0("Volcano Plot of Differential Expression\n", res[[i]]$contrast[1]),
       x = "Log2 Fold Change",
       #y = "-Log10 Adjusted P-value",
       y = "-Log10 P-value",
       color = "Significance")
print(plot)
  }
dev.off()







# interferon

pdf(file = paste0(OUT_DIR,"/inferferon_volcano.pdf")) 
  for (i in 1: length(res)){
# Input table

results = data.frame(
  adj.P.Val = res[[i]]$p_val_adj,
  logFC = res[[i]]$avg_log2FC,
  Gene = res[[i]]$gene,
  P.Val = res[[i]]$p_val
)


# Define thresholds (adjustable by the user)
sig_pval <- 0.05
lfc_threshold <- 0.1

# Allow the user to define an additional list of gene names to highlight (modify as needed)
extra_genes <- genes_in  # Replace with desired gene names

# Ensure gene names are available in the results
results$Gene <- results$Gene

# Define significance categories based on adjusted p-value and log2 fold change thresholds
results$Significance <- "NotSignificant"
results$Significance[results$adj.P.Val < sig_pval & results$logFC >= lfc_threshold] <- "Upregulated"
results$Significance[results$adj.P.Val < sig_pval & results$logFC <= -lfc_threshold] <- "Downregulated"
results$Significance <- factor(results$Significance, levels = c("Upregulated", "Downregulated", "NotSignificant"))

# Calculate -log10 of adjusted p-values for plotting
results$log10AdjP <- -log10(results$adj.P.Val)
results$log10P <- -log10(results$P.Val)
# Determine symmetrical x-axis limits based on maximum absolute log2 fold change (rounded up)
max_lfc <- ceiling(max(abs(results$logFC), na.rm = TRUE))
x_limits <- c(-max_lfc, max_lfc)

# Identify top 10 upregulated and top 10 downregulated proteins
top_up <- results %>% 
  dplyr::filter(Significance == "Upregulated") %>% 
  dplyr::arrange(-logFC) %>% 
  head(10)
top_down <- results %>% 
  dplyr::filter(Significance == "Downregulated") %>% 
  dplyr::arrange(logFC) %>% 
  head(10)

# Additional genes to highlight (if present in the results)
extra_highlight <- results %>% 
  dplyr::filter(Gene %in% extra_genes)

# Combine labels (avoid duplicates)
#labels_df <- dplyr::bind_rows(top_up, top_down, extra_highlight) %>% 
labels_df <- dplyr::bind_rows(extra_highlight) %>% 
  dplyr::distinct(Gene, .keep_all = TRUE)

# Define colors using the viridis palette for upregulated and downregulated proteins
up_color <- viridis(2)[2]
down_color <- viridis(2)[1]

# Generate the volcano plot
#ggplot(results, aes(x = logFC, y = log10AdjP)) +
plot = ggplot(results, aes(x = logFC, y = log10P)) +
  geom_point(aes(color = Significance), alpha = 0.6) +
  scale_color_manual(values = c("Upregulated" = up_color, 
                                "Downregulated" = down_color, 
                                "NotSignificant" = "grey50")) +
  geom_vline(xintercept = c(-lfc_threshold, lfc_threshold), linetype = "dashed", color = "red") +
  #geom_hline(yintercept = -log10(sig_pval), linetype = "dashed", color = "red") +
  geom_text_repel(data = labels_df, 
                  aes(label = Gene),
                  size = 3,
                  max.overlaps = Inf,
                  min.segment.length = 0
                  ) +
  coord_cartesian(xlim = x_limits) +
  theme_bw() +
  labs(title = paste0("Volcano Plot of Differential Expression\n", res[[i]]$contrast[1]),
       x = "Log2 Fold Change",
       #y = "-Log10 Adjusted P-value",
       y = "-Log10 P-value",
       color = "Significance")
print(plot)

  }
dev.off()


```




## Save data

```{r}
save.image("enrichment.RData")
```


